subsys/net/lib/sockets/getaddrinfo.c

```diff
@@ -107,14 +107,42 @@ static int exec_query(const char *host, int family,
 		      struct getaddrinfo_state *ai_state)
 {
 	enum dns_query_type qtype = DNS_QUERY_TYPE_A;
+	int st, ret;
 
-	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+	if (family == AF_INET6) {
 		qtype = DNS_QUERY_TYPE_AAAA;
 	}
 
-	return dns_get_addr_info(host, qtype, &ai_state->dns_id,
-				 dns_resolve_cb, ai_state,
-				 CONFIG_NET_SOCKETS_DNS_TIMEOUT);
+	ret = dns_get_addr_info(host, qtype, &ai_state->dns_id,
+				dns_resolve_cb, ai_state,
+				CONFIG_NET_SOCKETS_DNS_TIMEOUT);
+	if (ret == 0) {
+		/* If the DNS query for reason fails so that the
+		 * dns_resolve_cb() would not be called, then we want the
+		 * semaphore to timeout so that we will not hang forever.
+		 * So make the sem timeout longer than the DNS timeout so that
+		 * we do not need to start to cancel any pending DNS queries.
+		 */
+		ret = k_sem_take(&ai_state->sem, K_MSEC(CONFIG_NET_SOCKETS_DNS_TIMEOUT + 100));
+		if (ret == -EAGAIN) {
+			(void)dns_cancel_addr_info(ai_state->dns_id);
+			st = DNS_EAI_AGAIN;
+		} else {
+			st = ai_state->status;
+		}
+	} else if (ret == -EPFNOSUPPORT) {
+		/* If we are returned -EPFNOSUPPORT then that will indicate
+		 * wrong address family type queried. Check that and return
+		 * DNS_EAI_ADDRFAMILY and set errno to EINVAL.
+		 */
+		errno = EINVAL;
```

Comment:
I know it's a copy-paste, but `errno` only needs to be fixed for `DNS_EAI_SYSTEM`, perhaps a good occasion to fix that.

Reply:
Right, I'll remove this.